pulp_python/app/pypi/serializers.py

```diff
@@ -0,0 +1,90 @@
+import logging
+from gettext import gettext as _
+
+from rest_framework import serializers
+from pulp_python.app.tasks.upload import DIST_EXTENSIONS
+from pulpcore.plugin.models import Artifact
+from django.db.utils import IntegrityError
+
+log = logging.getLogger(__name__)
+
+
+class SummarySerializer(serializers.Serializer):
+    """
+    A Serializer for summary information of an index.
+    """
+
+    projects = serializers.IntegerField(help_text=_("Number of Python projects in index"))
+    releases = serializers.IntegerField(help_text=_("Number of Python distributions in index"))
+    files = serializers.IntegerField(help_text=_("Number of files for all distributions in index"))
+
+
+class PackageMetadataSerializer(serializers.Serializer):
+    """
+    A Serializer for a package's metadata.
+    """
+
+    last_serial = serializers.IntegerField(help_text=_("Cache value from last PyPI sync"))
+    info = serializers.JSONField(help_text=_("Core metadata of the package"))
+    releases = serializers.JSONField(help_text=_("List of all the releases of the package"))
+    urls = serializers.JSONField()
+
+
+class PackageUploadSerializer(serializers.Serializer):
+    """
+    A Serializer for Python packages being uploaded to the index.
+    """
+
+    content = serializers.FileField(
+        help_text=_("A Python package release file to upload to the index."),
+        required=True,
+        write_only=True,
+    )
+    action = serializers.CharField(
+        help_text=_("Defaults to `file_upload`, don't change it or request will fail!"),
+        default="file_upload",
+        source=":action"
+    )
+    sha256_digest = serializers.CharField(
+        help_text=_("SHA256 of package to validate upload integrity."),
+        required=False,
+        min_length=64,
+        max_length=64,
+    )
+
+    def validate(self, data):
+        """Validates the request."""
+        action = data.get("action")
+        if action != "file_upload":
+            raise serializers.ValidationError(
+                _("We do not support the :action {}".format(action))
+            )
+        file = data.get("content")
+        for ext, packagetype in DIST_EXTENSIONS.items():
+            if file.name.endswith(ext):
+                break
+        else:
+            raise serializers.ValidationError(_(
+                "Extension on {} is not a valid python extension "
+                "(.whl, .exe, .egg, .tar.gz, .tar.bz2, .zip)").format(file.name)
+            )
+        sha256 = data.get("sha256_digest")
+        digests = {"sha256": sha256} if sha256 else None
+        artifact = Artifact.init_and_validate(file, expected_digests=digests)
+        try:
+            artifact.save()
+        except IntegrityError:
+            artifact = Artifact.objects.get(sha256=artifact.sha256)
```

Comment:
This is inconsistent with Warehouse. The standard PyPI returns 403 on duplicate artifacts.

Reply:
Yes, inconsistent with Warehouse, but not so for Pulp. One of Pulp's key features is artifact de-duplication. If an artifact is shared among multiple content or a piece of content is contained in multiple repositories, Pulp only stores the one artifact or content unit once. So for pulp python the content unit is the Python package and its artifact is the tarball or wheel itself. Since Pulp can have multiple repositories if the package is already present in Pulp, but not in the repository, then it should be treated as adding that package to that repository.

Reply:
What if the artifact exists but the hash is different? It mustn't be possible to mutate the same dist. At least this case should error out.

Reply:
The `Artifact.init_and_validate` will throw an error if the digest is wrong, so the same dist can't be mutated. However, if the package and its artifact already exist and you try to upload the same package with a different artifact, Pulp will just use the present artifact instead. This could be considered incorrect behavior.

Reply:
> you try to upload the same package with a different artifact

With the same hash or a different one?

Reply:
Either or. Currently artifacts in Pulp are unique by their hash, but python packages in pulp_python are only unique by their filename. Since a goal of Pulp is to allow hosting multiple indexes, I think I'm going to need to change the global uniqueness constraint for python packages to be their hash, but still have the local uniqueness constraint for repositories be their filename.

Reply:
I've changed the uniqueness constraint for packages to be their sha256. If you upload a package with the same hash as the one already in Pulp, but not in the repository you will simply add that package to that repository. If you try to upload a package with the same name as another in Pulp but not in the repository and with a different hash then a new package will be created and added to the repository.  This means there can be two packages with the same name but different digests in Pulp, however they both can't be in the same repository. The sha256 of the package's metadata and its artifact are now linked and can't be mismatched.

Reply:
> This means there can be two packages with the same name but different digests in Pulp, however they both can't be in the same repository. 

Awesome, this makes perfect sense 👍🏻 thx!